core/src/OpenACC/Kokkos_OpenACC_ParallelFor_Team.hpp

```diff
@@ -0,0 +1,192 @@
+
+//----------------------------------------------------------------------------
+//----------------------------------------------------------------------------
+// Hierarchical Parallelism -> Team level implementation
+template <class FunctorType, class... Properties>
+class Kokkos::Impl::ParallelFor<FunctorType, Kokkos::TeamPolicy<Properties...>,
+                                Kokkos::Experimental::OpenACC> {
+ private:
+  using Policy = Kokkos::Impl::TeamPolicyInternal<Kokkos::Experimental::OpenACC,
+                                                  Properties...>;
+  Kokkos::Experimental::Impl::FunctorAdapter<FunctorType, Policy> m_functor;
+  using Member = typename Policy::member_type;
+
+  const Policy m_policy;
+
+ public:
+  inline void execute() const {
+    auto league_size   = m_policy.league_size();
+    auto team_size     = m_policy.team_size();
+    auto vector_length = m_policy.impl_vector_length();
+
+    auto const a_functor(m_functor);
+
+#ifdef KOKKOS_ENABLE_OPENACC_COLLAPSE_HIERARCHICAL_CONSTRUCTS
+#pragma acc parallel loop gang vector num_gangs(league_size) \
+    vector_length(team_size* vector_length) copyin(a_functor)
+    for (int i = 0; i < league_size * team_size * vector_length; i++) {
+      int league_id = i / (team_size * vector_length);
+      typename Policy::member_type team(league_id, league_size, team_size,
+                                        vector_length);
+      a_functor(team);
+    }
+#else
+#pragma acc parallel loop gang num_gangs(league_size) num_workers(team_size) \
+    vector_length(vector_length) copyin(a_functor)
+    for (int i = 0; i < league_size; i++) {
+      int league_id = i;
+      typename Policy::member_type team(league_id, league_size, team_size,
+                                        vector_length);
+      a_functor(team);
+    }
+#endif
+  }
+
+  inline ParallelFor(const FunctorType& arg_functor, Policy arg_policy)
+      : m_functor(arg_functor), m_policy(arg_policy) {}
+};
+
+// Hierarchical Parallelism -> Team thread level implementation
+namespace Kokkos {
+
+#ifdef KOKKOS_ENABLE_OPENACC_COLLAPSE_HIERARCHICAL_CONSTRUCTS
+#pragma acc routine seq
+#else
+#pragma acc routine worker
+#endif
+template <typename iType, class Lambda>
+KOKKOS_INLINE_FUNCTION void parallel_for(
+    const Impl::TeamThreadRangeBoundariesStruct<iType, Impl::OpenACCTeamMember>&
+        loop_boundaries,
+    const Lambda& lambda) {
+#ifdef KOKKOS_ENABLE_OPENACC_COLLAPSE_HIERARCHICAL_CONSTRUCTS
+  iType j_start =
+      loop_boundaries.team.team_rank() / loop_boundaries.team.vector_length();
+  iType j_end  = loop_boundaries.end;
+  iType j_step = loop_boundaries.team.team_size();
+  if (j_start >= loop_boundaries.start) {
+#pragma acc loop seq
+    for (iType j = j_start; j < j_end; j += j_step) {
+      lambda(j);
+    }
+  }
+#else
+#pragma acc loop worker
+  for (iType j = loop_boundaries.start; j < loop_boundaries.end; j++) {
+    lambda(j);
+  }
+#endif
+}
```

Comment:
As the only repeating lines are the function interface and the braces, I would prefer to have two separate functions instead of one with `#ifdef` inside ... I think it would be easier to read

Reply:
Refactored as suggested.